rfcs/2021-06-15-shipping-blade.md

```diff
@@ -0,0 +1,438 @@
+    - [Conclusion](#conclusion-2)
+  - [Package consumption information in `package.json`](#package-consumption-information-in-packagejson)
+    - [`exports` map field](#exports-map-field)
+    - [`main`, `module`, `browser`, `react-native` fields in `package.json`](#main-module-browser-react-native-fields-in-packagejson)
+    - [Conclusion](#conclusion-3)
+- [Drawbacks/Constraints](#drawbacksconstraints)
+- [Alternatives](#alternatives)
+- [Adoption strategy](#adoption-strategy)
+- [Open Questions](#open-questions)
+- [References](#references)
+
+# Summary
+
+We can't ship the raw code to the consumers so we need to ship/publish blade. But to make it publishable there are certain things we need to keep in mind and answer relevant questions.
+
+# Basic Example
+We should be able to install and use blade in our code without any overhead.
+```js
+import { Button, Text } from '@razorpay/blade/components'
+import { overrideTheme, paymentTheme } from '@razorpay/blade/theme'
+import { getColorScheme, useColorScheme } from '@razorpay/blade/utils'
+```
+
+# Motivation
+- We can't ship raw source code to our consumers since we are not aware about the build tool chains they use, what config do they support.
+- We need to prevent access to certain parts of the codebase. There might be certain files/functionality which are just used for internal purpose and we don't want to expose them to our consumers. So we want to publish few things and not publish few.
+
+In the end we should be able to ship the code in such a way that:
+* It doesn't increases the bundle size of our consumer apps.
+* There's no compilation issue on their end.
+* We restrict them to only use what is useful for them and hide away unnecessary underlying details.
+
+# Detailed Design
+This problem has multiple pieces to it and in order to come up with a solution I have tried to break it down into questions that we must ask to ourselves and then conclude every question.
+
+Here's the visual representation for decision making of bundling. 
+> You can look at the actual source [file of the diagram here](https://drive.google.com/file/d/1AM0ryQCXMbfP4z2PUWpt86KUU5iE8rvR/view?usp=sharing) if you can't see things properly in attached image
+
+![Blade Bundling](./images/bundling-blade-guideline.png)
+
+## What do we want to actually ship?
+* components
+* tokens/themes(need to decide on the naming)
+* types
+* utils
+
+## How will the consumers import things from blade?
+There are multiple ways that consumers can import components, themes, types, utils from blade. Let's look into it one by one
+
+### 1. `import` individual components
+```js
+/* multi-package */
+import Button from '@razorpay-blade/Button' 
+import Text from '@razorpay-blade/Text'
+import overrideTheme from '@razorpay-blade/overrideTheme'
+import paymentTheme from '@razorpay-blade/paymentTheme'
+import getColorScheme from '@razorpay-blade/getColorScheme'
+import useColorScheme from '@razorpay-blade/useColorScheme'
```

Comment:
In multi-package, these functions would be part of one package. `@razorpay-blade/theme`